rfcs/text/0000-metadata-integration-test.md

```diff
@@ -0,0 +1,106 @@
+1. The agility of the metadata and ingest teams is hampered
+
+To solve these issues we need to
+
+1. Make the development environment more stable
+1. Make updates to software and schema smaller and more frequent
+1. Make failures observable
+1. Make systems downstream of ingest more agnostic to schema or fail transparently
+
+## Detailed Design
+
+Below we describe a design that will transition DCP from the current "waterfall" software/metadata integration model to a more observable and interactive integration model, informed by integration tests that display the true compatibility state of the system.
+
+Setting this plan will also better define the interaction between the metadata team and the rest of the DCP. Processes and responsibilities for resolving any integration issues that arise will be clear.
+
+### Design Alternative #1
+
+Software development proceeds in the dev branch of each component independently. Changes are merged into integration branch of each component only through PRs which cause system-wide integration tests to run. PRs are merged automatically if tests pass (but can be reverted if downstream breakage occurs). PRs may not be merged if tests do not pass.
+
+![Option 1](../images/0000-metadata-integration-test-opt1.png)
+
+*Here DSS Dev is shown as an example of other components following the same process.*
+
+Data with the new schema version should not be uploaded into the `integration` or subsequent deployment stages until the test passes and the `master` branch is promoted to `integration`.
+
+#### Pros
+
+* "Test then merge"
+* PR status will not require wranglers to know about downstream systems
+* Small non-breaking changes can be merged quickly (out-of-order) and with confidence
+* Greatly reduces probability of deployment stages past integration breaking
+* More certainty for developers on behavior of the system
+* Issues caused by metadata changes are surfaced early
+* All DCP projects are integrated in the same way
+
+#### Cons
+
+* May be difficult to revert test versions of components and metadata (not all infra deployments are easily reversible)
+* Requires "serializing" the dcp-wide integration test so that one PR can run after another
+
+### Design Alternative #2 
+
+Software development proceeds in the dev branch of each component independently.
+When a metadata schema version has passed unit tests in `dev` and its committers wish to promote it, they make a PR against their integration branch.
+The PR triggers a dcp-wide test in the integration environment.
+Metadata schema changes can be merged from `dev` to `integration` only if the integration test suite passes.
+
+![Option 2](../images/0000-metadata-integration-test-opt2.png)
+
+Data with the new schema version should not be uploaded into the `integration` or subsequent deployment stages until the test passes and the `master` branch is promoted to `integration`.
```

Comment:
_**should not** be uploaded_ or **must not**? See [RFC2119](https://www.ietf.org/rfc/rfc2119.txt):

>4. SHOULD NOT   This phrase, or the phrase "NOT RECOMMENDED" mean that
   there may exist valid reasons in particular circumstances when the
   particular behavior is acceptable or even useful, but the full
   implications should be understood and the case carefully weighed
   before implementing any behavior described with this label.

Reply:
I think this one can be a "should not."

Reply:
If that is the case, then the _valid reasons in particular circumstances_ could be documented; otherwise, there's a tendency to interpret *SHOULD NOT* as *OPTIONAL*.